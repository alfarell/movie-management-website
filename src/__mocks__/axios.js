export default {
    all: jest.fn().mockResolvedValue([]),
    get: jest.fn().mockResolvedValue({ data: {} })
};