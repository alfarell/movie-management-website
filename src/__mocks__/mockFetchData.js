import Axios from 'axios';

const mockFetchData = async (mockData) => {
    await Axios.get.mockResolvedValueOnce({
        data: mockData,
    });
}

const mockFetchAllData = async (mockAllData) => {
    await Axios.all.mockResolvedValueOnce(mockAllData);
}

export { mockFetchData, mockFetchAllData };