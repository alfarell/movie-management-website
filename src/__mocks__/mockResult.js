export const mockGenreList = {
    genres: [
        {
            id: 1,
            name: 'Fantasy'
        },
        {
            id: 2,
            name: 'Sci-Fi'
        }
    ]
}

export const mockMovieList = {
    results: [
        {
            id: 1,
            title: 'Test Movie List',
            poster_path: 'test_poster',
            average_score: 10
        },
        {
            id: 2,
            title: 'Test Movie List 2',
            poster_path: null,
            average_score: 10
        },
    ]
}

export const mockLoadMoreMovie = {
    results: [
        {
            id: 3,
            title: 'Test Load More Movie',
            poster_path: null,
            average_score: 10
        },
    ]
}

export const mockMovieDetail = {
    id: 1,
    title: 'Test Title',
    tagline: 'Test Tagline',
    poster_path: '/test_poster',
    backdrop_path: '/test_backdrop',
    vote_average: 10,
    overview: 'Test Overview',
    genres: [{ id: 1, name: 'Test Genre' }],
    release_date: '2020-01-01',
    vote_count: 100
}

export const mockMovieCast = {
    cast: [
        {
            id: 1,
            name: 'Test Cast',
            profile_path: 'profile_test'
        },
        {
            id: 2,
            name: 'Test Cast 2',
            profile_path: null
        }
    ]
}