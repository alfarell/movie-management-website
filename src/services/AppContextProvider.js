import React, { createContext, useState, useEffect } from 'react';
import _ from 'loadsh';
import Axios from 'axios';


export const MovieContext = createContext();
export const UserContext = createContext();

const AppContextProvider = ({ children }) => {
    const [movieList, setMovieList] = useState([]);
    const [genreList, setGenreList] = useState([]);
    const [selectedGenre, setSelectedGenre] = useState(null);
    const [sortOption, setSortOption] = useState('popularity.desc')
    const [pagination, setPagination] = useState(1);


    useEffect(() => {
        Axios.get(`${process.env.REACT_APP_BASE_URL}/genre/list?api_key=${process.env.REACT_APP_BASE_API_KEY}&language=en-US`)
            .then(res => {
                setGenreList(res.data.genres);
            })
            .catch(err => {
                console.log(err);
            });
    }, [genreList]);

    useEffect(() => {
        const discovers = {
            genre: selectedGenre ? selectedGenre.id : '',
            sort: sortOption,
        };

        setPagination(1);
        fetchMovie(discovers);
    }, [selectedGenre, sortOption]);

    const fetchMovie = ({ genre, sort }) => {
        Axios.get(`${process.env.REACT_APP_BASE_URL}/discover/movie?api_key=${process.env.REACT_APP_BASE_API_KEY}&language=en-US&include_adult=false&include_video=false&page=${pagination}&with_genres=${genre}&sort_by=${sort}`)
            .then(res => {
                setMovieList([...res.data.results])
            })
            .catch(err => {
                console.log(err);
            });
    }

    const handleLoadMore = () => {
        setPagination(pagination + 1);

        const genre = selectedGenre ? selectedGenre.id : '';
        const sort = sortOption;
        let page = pagination + 1

        Axios.get(`${process.env.REACT_APP_BASE_URL}/discover/movie?api_key=${process.env.REACT_APP_BASE_API_KEY}&language=en-US&include_adult=false&include_video=false&page=${page}&with_genres=${genre}&sort_by=${sort}`)
            .then(res => {
                setMovieList([...movieList, ...res.data.results])
            })
            .catch(err => {
                console.log(err);
            });
    }

    const addFavoriteMovie = (favoritedMovie) => {
        let currentMovie = JSON.parse(localStorage.getItem('favorited'));
        let addFavorited;

        if (currentMovie) {
            let check = _.find(currentMovie, favoritedMovie);
            if (check) {
                addFavorited = currentMovie.filter(movie => {
                    return movie.id !== favoritedMovie.id
                });
            } else {
                addFavorited = [...currentMovie, favoritedMovie]
            }
        } else {
            addFavorited = [favoritedMovie];
        }

        localStorage.setItem('favorited', JSON.stringify(addFavorited));
    }

    return (
        <MovieContext.Provider value={{
            movieList,
            setMovieList,
            genreList,
            setGenreList,
            selectedGenre,
            setSelectedGenre,
            sortOption,
            setSortOption,
            pagination,
            setPagination,
            handleLoadMore,
            addFavoriteMovie
        }}>
            <UserContext.Provider value={{ test: 'test' }}>
                {children}
            </UserContext.Provider>
        </MovieContext.Provider>
    );
};

export default AppContextProvider;
