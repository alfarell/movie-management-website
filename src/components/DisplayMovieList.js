import React, { useContext, Fragment } from 'react';
import { Row, Spin, Col, Button, Alert } from 'antd';
import { MovieContext } from '../services/AppContextProvider';
import CardContent from './CardContent';


const RenderMovieList = ({ movieList, isLoading, error }) => {
    if (isLoading.loading && isLoading.loader === 'movie-list') {
        return <Spin tip='Loading...' size='large' />;
    }
    if (error.status && error.error === 'fetch-movie-error') {
        return (
            <Alert
                description='Some Error is Occured, Please check your internet connection and refresh the page'
                type='error'
                showIcon
            />
        );
    }

    return movieList.map(movie => {
        return (
            <Col key={movie.id} xs={11} sm={7} md={6} lg={6} xl={4} xxl={3} >
                <CardContent data={movie} />
            </Col>
        )
    });
};

const DisplayMovieList = () => {
    const { movieList, handleLoadMore, isLoading, error } = useContext(MovieContext);

    return (
        <Fragment>
            <Row gutter={[10, 10]} justify='center'>
                <RenderMovieList {...{ movieList, isLoading, error }} />
            </Row>

            {movieList.length === 0
                ? null
                : <Fragment>
                    <Button
                        type='primary'
                        loading={isLoading.loading && isLoading.loader === 'load-more-movie'}
                        onClick={handleLoadMore}
                    >
                        Load more
                    </Button>
                    <Alert
                        message='Error'
                        description='Failed to load more movie, please check your internet connection and try again'
                        type='error'
                        showIcon
                        style={{ display: error.status && error.error === 'load-more-movie-error' ? 'block' : 'none' }}
                    />
                </Fragment>
            }
        </Fragment>
    );
};

export default DisplayMovieList;
