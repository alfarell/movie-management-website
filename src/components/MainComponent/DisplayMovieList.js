import React, { useContext, Fragment } from 'react';
import { Row, Spin, Col, Button, Alert } from 'antd';
import { MovieContext } from '../../services/AppContextProvider';
import CardContent from './CardContent';


const DisplayMovieList = () => {
    const { movieList, setPagination, loading, error } = useContext(MovieContext);

    return (
        <Fragment>
            <Row gutter={[10, 10]} justify='center'>
                {movieList.map(movie => {
                    return (
                        <Col key={movie.id} xs={11} sm={7} md={6} lg={6} xl={4} xxl={3} data-testid='movie-list'>
                            <CardContent data={movie} />
                        </Col>
                    )
                })}
                {loading && <Spin tip='Loading...' size='large' />}
                {error.status && (<Alert
                    description='Some Error is Occured, Please check your internet connection and try again or refresh the page'
                    type='error'
                    showIcon
                    data-testid='movielist-error-message'
                />)}
            </Row>

            {!loading && movieList.length > 0 &&
                (<Button
                    type='primary'
                    loading={loading}
                    onClick={() => setPagination(pagination => pagination + 1)}
                >
                    Load more
                </Button>)
            }
        </Fragment>
    );
};

export default DisplayMovieList;
