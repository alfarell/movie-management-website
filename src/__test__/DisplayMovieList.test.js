import React from 'react';
import Axios from 'axios';
import { render, waitForElement, act, fireEvent } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import AppContextProvider from '../services/AppContextProvider';
import DisplayMovieList from '../components/MainComponent/DisplayMovieList';
import { mockFetchData } from '../__mocks__/mockFetchData';
import { mockGenreList, mockMovieList, mockLoadMoreMovie } from '../__mocks__/mockResult';


Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

describe('Display Movie List', () => {
    afterEach(() => {
        Axios.get.mockClear();
    });

    it('render Display Movie List', async () => {
        await mockFetchData(mockGenreList);
        await mockFetchData(mockMovieList);

        const { container, asFragment, getByText } = render(
            <AppContextProvider>
                <Router>
                    <DisplayMovieList />
                </Router>
            </AppContextProvider>
        );

        expect(asFragment).toMatchSnapshot();
        expect(getByText('Loading...')).toBeInTheDocument();
        expect(container).toHaveTextContent('Loading...');

        const movieList = await waitForElement(() => getByText('Test Movie List'));

        expect(movieList).toBeInTheDocument();
    });

    it('should show more movie list when user load more movie', async () => {
        await mockFetchData(mockGenreList);
        await mockFetchData(mockMovieList);

        const { getByText } = render(
            <AppContextProvider>
                <Router>
                    <DisplayMovieList />
                </Router>
            </AppContextProvider>
        );

        expect(getByText('Loading...')).toBeInTheDocument();

        const movieList = await waitForElement(() => getByText('Test Movie List'));
        expect(movieList).toBeInTheDocument();

        await act(async () => {
            await fireEvent.click(getByText('Load more'));
            await mockFetchData(mockLoadMoreMovie);
        });

        const moreMovieList = await waitForElement(() => getByText('Test Load More Movie'));
        expect(moreMovieList).toBeInTheDocument();
    });

    it('should show error message when failed to get data', async () => {
        await mockFetchData(mockGenreList);
        await Axios.get.mockRejectedValueOnce(new Error('error'));

        const { container, getByText, getByTestId } = render(
            <AppContextProvider>
                <Router>
                    <DisplayMovieList />
                </Router>
            </AppContextProvider>
        );

        expect(getByText('Loading...')).toBeInTheDocument();
        expect(container).toHaveTextContent('Loading...');

        const errorMessage = await waitForElement(() => getByTestId('movielist-error-message'));

        expect(errorMessage).toBeInTheDocument();
    });
});
