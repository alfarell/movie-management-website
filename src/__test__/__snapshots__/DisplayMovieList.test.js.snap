// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Movie List render Display Movie List 1`] = `[Function]`;
